Prédiction de l’effort de développement d’un projet logiciel
             EL ARGOUBI CHAIMAA
  


notre but c'est de concevoir un modèle intelligent qui permet de prédire l’effort de développement d’un projet logiciel

# IMPORTATION DES DONNEES

Dans un premier temps, on charge les données.

In [141]:
import pandas as pd
df=pd.read_csv("zia.csv",sep = ';')
df

storyPoint  velocity  Effort
0          156       2.7      63
1          202       2.5      92
2          173       3.3      56
3          331       3.8      86
4          124       4.2      32
5          339       3.6      91
6           97       3.4      35
7          257       3.0      93
8           84       2.4      36
9          211       3.2      62
10         131       3.2      45
11         112       2.9      37
12         101       2.9      32
13          74       2.9      30
14          62       2.9      21
15         289       2.8     112
16         113       2.8      39
17         141       2.8      52
18         213       2.8      80
19         137       2.7      56
20          91       2.7      35

In [121]:
Team_Size = [5, 5,5,5,5, 7, 5, 5,5,5,5,5,5,7,5,5,5,5,5,5,5]

df['Team Size'] = Team_Size
 
# Observe the result
print(df)


    storyPoint  velocity  Effort  Team Size
0          156       2.7      63          5
1          202       2.5      92          5
2          173       3.3      56          5
3          331       3.8      86          5
4          124       4.2      32          5
5          339       3.6      91          7
6           97       3.4      35          5
7          257       3.0      93          5
8           84       2.4      36          5
9          211       3.2      62          5
10         131       3.2      45          5
11         112       2.9      37          5
12         101       2.9      32          5
13          74       2.9      30          7
14          62       2.9      21          5
15         289       2.8     112          5
16         113       2.8      39          5
17         141       2.8      52          5
18         213       2.8      80          5
19         137       2.7      56          5
20          91       2.7      35          5


Notre dataset est constituée de 4 colonnes Story Point qui represente la taille du projet en story point,Velocity represente la quantité (en story point) que l’équipe
peut développer pendant une période,Team Size represente l’effectif de l’équipe de développement et finalement on a la colonne Effort qu'on veut prédire ses valeurs

# PREPROCESSING DES DONNEES
 


 Avant de passer à la phase de classification, il faut nettoyer nos données ! Cette phase, appelée preprocessing en anglais, est essentielle et sur elle repose 80% de la réussite de tel ou tel algorithme.
 Donc dans un premier temps on va vérifier si il y a des valeurs null dans notre table 

In [124]:
print(df.isnull().sum())

storyPoint    0
velocity      0
Effort        0
Team Size     0
dtype: int64


On remarque trés bien qu'il n'ya aucune valeur nule dans notre table,donc on peut maintenant séparer nos données en un jeu d’entraînement et un jeu de test, afin d’évaluer la qualité de notre modèle, puis on peut directement entraîner le modèle et effectuer des prédictions sur le jeu de test. 

In [129]:
x=df.drop(columns=["Effort"])
y=df["Effort"]

In [126]:
x

storyPoint  velocity  Team Size
0          156       2.7          5
1          202       2.5          5
2          173       3.3          5
3          331       3.8          5
4          124       4.2          5
5          339       3.6          7
6           97       3.4          5
7          257       3.0          5
8           84       2.4          5
9          211       3.2          5
10         131       3.2          5
11         112       2.9          5
12         101       2.9          5
13          74       2.9          7
14          62       2.9          5
15         289       2.8          5
16         113       2.8          5
17         141       2.8          5
18         213       2.8          5
19         137       2.7          5
20          91       2.7          5

In [142]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
x_train , x_test,y_train , y_test=train_test_split(x,y,test_size=0.3,random_state=1)
from sklearn import tree
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
model = KNeighborsRegressor(n_neighbors=3)
model2=tree.DecisionTreeRegressor()
model3=LinearRegression()

Alors on va évaluer notre modèle à l’aide des classifieurs suivants : KNN,Arbres de décisions et Régression Lineaire.
Avec un score  assez élevé et une matrice de confusion concluante, on peut déduire quel est e meilleur classifieur


# ENTRAINEMENT DU MODELE

TECHNOLOGIE 1 : KNeighborsRegressor

In [138]:
#KNeighborsRegressor
print("RESULTATS DONNES PAR LE MODELE KNeighborsRegressor : ")
model.fit(x_train,y_train)
a=model.score(x_train,y_train)
print("accuracy sur train :",a,"%")
a2=model.score(x_test,y_test)
print("accuracy sur test :",a2,"%")
y_pred=model.predict(x_test)
#calcule des indicateurs MMRE et Pred(25%)
ABE=abs(y_test-y_pred)
MRE=ABE/y_test
import statistics
MMRE= statistics.mean(MRE)
a=[]
for i in MRE:
    if i<=0.25:
        a.append(1)
    else:
        a.append(0)

Pred=sum(a)/len(a)   
print("l'indicateur MMRE de ce model est :",MMRE ,"%")
print("l'indicateur PRED(25%) de ce model est :",Pred ,"%")

RESULTATS DONNES PAR LE MODELE KNeighborsRegressor : 
accuracy sur train : 0.8464483418697295 %
accuracy sur test : 0.7968316716411592 %
l'indicateur MMRE de ce model est : 0.18363155240231652 %
l'indicateur PRED(25%) de ce model est : 0.7142857142857143 %


TECHNOLOGIE 2 : DecisionTreeRegressor

In [139]:
#DecisionTreeRegressor
print("RESULTATS DONNES PAR LE MODELE DecisionTreeRegressor : ")
model2.fit(x_train,y_train)
a=model2.score(x_train,y_train)
print("accuracy sur train :",a,"%")
a2=model2.score(x_test,y_test)
print("accuracy sur test :",a2,"%")
y_pred=model2.predict(x_test)
#calcule des indicateurs MMRE et Pred(25%)
ABE=abs(y_test-y_pred)
MRE=ABE/y_test
import statistics
MMRE= statistics.mean(MRE)
a=[]
for i in MRE:
    if i<=0.25:
        a.append(1)
    else:
        a.append(0)

Pred=sum(a)/len(a)   
print("l'indicateur MMRE de ce model est :",MMRE ,"%")
print("l'indicateur PRED(25%) de ce model est :",Pred ,"%")

RESULTATS DONNES PAR LE MODELE DecisionTreeRegressor : 
accuracy sur train : 1.0 %
accuracy sur test : 0.8746427891026862 %
l'indicateur MMRE de ce model est : 0.16842905617822562 %
l'indicateur PRED(25%) de ce model est : 0.8571428571428571 %


TECHNOLOGIE 3 : Linear Regression

In [140]:
#Linear Regression
print("RESULTATS DONNES PAR LE MODELE lINEAR REGRESSION : ")
model3.fit(x_train,y_train)
a=model3.score(x_train,y_train)
print("accuracy sur train :",a,"%")
a2=model3.score(x_test,y_test)
print("accuracy sur test :",a2,"%")
y_pred=model3.predict(x_test)
#calcule des indicateurs MRE et ABE
ABE=abs(y_test-y_pred)
MRE=ABE/y_test
import statistics
MMRE= statistics.mean(MRE)
a=[]
for i in MRE:
    if i<=0.25:
        a.append(1)
    else:
        a.append(0)

Pred=sum(a)/len(a)   
print("l'indicateur MMRE de ce model est :",MMRE ,"%")
print("l'indicateur PRED(25%) de ce model est :",Pred ,"%")



RESULTATS DONNES PAR LE MODELE lINEAR REGRESSION : 
accuracy sur train : 0.9218128883532224 %
accuracy sur test : 0.9374300534761124 %
l'indicateur MMRE de ce model est : 0.11229893045123487 %
l'indicateur PRED(25%) de ce model est : 0.8571428571428571 %


Maintenant on va essayer d'utiliser l'une des techniques permettant d’améliorer la performance du modèle, c'est la feature selection

Pour ce faire j'ai utiliser la fonction suivante "trouver_donnees importantes" qui va nous permettre de sélectionner
les fonctionnalités qui contribuent le plus à notre variable de prédiction.et ceci en utilisant le Sélecteur de fonction VarianceThreshold(threshold) qui supprime toutes les fonctions à variance inférieur au seuil de variance'threshold'.

In [143]:
#accuracy apres l'utilisation du feature selection
from sklearn.feature_selection import VarianceThreshold
def trouver_donnees_importantes(x_train,TR):
    Filtre=VarianceThreshold(threshold=TR)
    Donnees_Filtrees=Filtre.fit_transform(x_train)
    FS=list(Filtre.get_support());
    X=[]
    for i in range(0,len(FS)):
        if FS[i]==True:
            X+=[i]
    return tuple(X)


In [144]:
print(x_train.var(axis=0))

storyPoint    5771.670330
velocity         0.229890
Team Size        0.527473
dtype: float64


In [145]:
x_test.var(axis=0)

storyPoint    9674.809524
velocity         0.141429
Team Size        0.000000
dtype: float64

In [146]:
x_train.shape

(14, 3)

In [147]:
x2=trouver_donnees_importantes(x_train,0.4)
x1_train=x_train.to_numpy()
x3_train=x1_train[:,x2]
x3_train.shape

(14, 2)

In [148]:
y_train.shape


(14,)

Résultats apres utilisation de la technique Feature Selection:

TECHNOLOGIE 1: KNeighborsRegressor

In [149]:
#KNeighborsRegressor
model.fit(x3_train,y_train)
a=model.score(x3_train,y_train)
print("accuracy score sur train :",a,"%")
x1_test=x_test.to_numpy()
x3_test=x1_test[:,x2]
a2=model.score(x3_test,y_test)
print("accuracy score sur test :",a2,"%")
y_pred=model.predict(x3_test)
#calcule des indicateurs MMRE et Pred(25%)
ABE=abs(y_test-y_pred)
MRE=ABE/y_test
import statistics
MMRE= statistics.mean(MRE)
a=[]
for i in MRE:
    if i<=0.25:
        a.append(1)
    else:
        a.append(0)

Pred=sum(a)/len(a)   
print("l'indicateur MMRE de ce model est :",MMRE ,"%")
print("l'indicateur PRED(25%) de ce model est :",Pred ,"%")


accuracy score sur train : 0.8464483418697295 %
accuracy score sur test : 0.7968316716411592 %
l'indicateur MMRE de ce model est : 0.18363155240231652 %
l'indicateur PRED(25%) de ce model est : 0.7142857142857143 %


TECHNOLOGIE 2: DecisionTreeRegressor

In [154]:
#DecisionTreeRegressor 
model2.fit(x3_train,y_train)
a=model2.score(x3_train,y_train)
print("accuracy score sur train :",a,"%")
x1_test=x_test.to_numpy()
x3_test=x1_test[:,x2]
a2=model2.score(x3_test,y_test)
print("accuracy score sur test :",a2,"%")
y_pred=model2.predict(x3_test)
#calcule des indicateurs MMRE et Pred(25%)
ABE=abs(y_test-y_pred)
MRE=ABE/y_test
import statistics
MMRE= statistics.mean(MRE)
a=[]
for i in MRE:
    if i<=0.25:
        a.append(1)
    else:
        a.append(0)

Pred=sum(a)/len(a)   
print("l'indicateur MMRE de ce model est :",MMRE ,"%")
print("l'indicateur PRED(25%) de ce model est :",Pred ,"%")


accuracy score sur train : 1.0 %
accuracy score sur test : 0.8446370737283292 %
l'indicateur MMRE de ce model est : 0.17914334189251133 %
l'indicateur PRED(25%) de ce model est : 0.8571428571428571 %


TECHNOLOGIE 3: Linear Regression 

In [155]:
#Linear Regression 
model3.fit(x3_train,y_train)
a=model3.score(x3_train,y_train)
print("accuracy score sur train :",a)
x1_test=x_test.to_numpy()
x3_test=x1_test[:,x2]
a2=model3.score(x3_test,y_test)
print("accuracy score sur test :",a2)
y_pred=model3.predict(x3_test)
#calcule des indicateurs MMRE et Pred(25%)
ABE=abs(y_test-y_pred)
MRE=ABE/y_test
import statistics
MMRE= statistics.mean(MRE)
a=[]
for i in MRE:
    if i<=0.25:
        a.append(1)
    else:
        a.append(0)

Pred=sum(a)/len(a)   
print("l'indicateur MMRE de ce model est :",MMRE ,"%")
print("l'indicateur PRED(25%) de ce model est :",Pred ,"%")


accuracy score sur train : 0.8320789595026039
accuracy score sur test : 0.8540371151022709
l'indicateur MMRE de ce model est : 0.14295503251606842 %
l'indicateur PRED(25%) de ce model est : 0.8571428571428571 %


# Conclusion


D'aprés les résultats obtenus on remarque que les valeurs de l'accuracy sur test et de l'indicateur Pred(25%) données par  regression lineaire et decision tree apres utilisation de la feature selection  sont presque les memes, mais avant utilisation de feature selection, la regression lineaire donne une valeure de l'accuracy(0.94%) plus élevée que celle donnée par les autres techniques Decision Tree(0.87%) et KNN(0.79%).
On remarque aussi que la regression lineaire donne une meilleur valeur de l'indicateur MMRE(0.11% avant feature selection et 0.14% apres) en comparaison avec les autres techniques.
En faisant la comparaison entre les valeurs données par Regression Lineaire avant et apres utilisation de feature selection 
on peut voir clairement que c'avant que la regression lineaire donne des meilleurs valeurs .
Finalement, on peut déduire que la regression lineaire est le meilleur classifieur pour ce problème